refactor 'Contribute Code'

zhangyangyu opened the pull request:
<!-- Thank you for contributing to TiDB Development Guide!  -->
<!-- Please follow the PR title format:                     -->
<!--    "section: what's changed"                           -->

### What issue does this PR solve?

<!-- only need to keep one of the following lines -->

- close #129 

There are some information I think we need to give in this chapter but we don't have them yet:
1. still no clear TiDB branch strategy.
2. still no commit message rule consensus
3. still no TiDB roadmap or feature map
4. still no experts list 
5. still no clear label strategy

xxchan commented on src/contribute-to-tidb/contribute-code.md:
```diff
@@ -1,88 +1,50 @@
+## Before contributing
 
-Please read this document carefully before starting to work on a code contribution. Follow the process and guidelines explained below. Contributing to TiDB does _not_ start with opening a pull request. We expect contributors to reach out to us first to discuss the overall approach together. Without consensus with the TiDB committers, contributions might require substantial rework or will not be r ...
+Contributing to TiDB does *not* start with opening a pull request. We expect contributors to reach out to us first to discuss the overall approach together. Without consensus with the TiDB committers, contributions might require substantial rework or will not be reviewed. So please [create a GitHub issue](open-an-issue.md) or create a topic on the [internal.tidb.io](https://internals.tidb.io) and ...
```
```suggestion
Contributing to TiDB does *not* start with opening a pull request. We expect contributors to reach out to us first to discuss the overall approach together. Without consensus with the TiDB committers, contributions might require substantial rework or will not be reviewed. So please [create a GitHub issue](open-an-issue.md) or discuss under an existing issue or create a topic on the [internal.tidb.io](https://internals.tidb.io) and reach consensus.
```

tisonkun commented on src/contribute-to-tidb/contribute-code.md:
```diff
@@ -1,88 +1,53 @@
-2. Implement. Implement the change according to the approach agreed upon in the discussion.
-3. Review. Open a pull request and work with the reviewer.
-4. Merge. A committer of TiDB checks if the contribution fulfills the requirements and merges the code to the codebase.
+  1. Assign the issue to yourself via [\assign](https://prow.tidb.io/command-help?repo=pingcap%2Ftidb#assign). This lets other contributors know you are working on the issue so they won't make duplicate efforts.
+  2. Follow the [GitHub workflow](https://guides.github.com/introduction/flow/), commit code changes in your own git repository branch and open a pull request for code review.
+  3. Make sure the continuous integration checks on your pull request are green (i.e. successful).
+  4. Review and address [comments on your pull request](https://devguide.python.org/pullrequest/). If your pull request becomes outdated with the target branch, you need to [rebase your pull request](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request#perform-a-rebase) to keep it up to date.
+  5. When your pull request gets enough approvals(the default number is 2) and all other requirements are met, it will be merged.
+  6. Handle regressions introduced by your change. Although committers bear the main responsibility to fix regressions, it's quite nice for you to handle it(reverting the change or sending fixes).
```
Several comments:

1. I don't think we should link to Python Dev Guide here. If we want any of its content, write our own.
2. Because of we use "squash and merge" we don't have to rebase on master but often just merge master to catch up change.

zhangyangyu replied:
1. It's my mistake. Seems I paste a wrong link, it should be a link to a official GitHub guide.
2. Git practise could vary between people. merge and rebase should both work here.

tisonkun commented on src/contribute-to-tidb/contribute-code.md:
```diff
@@ -1,88 +1,53 @@
+When creating a pull request for submission, there are several things that you should consider to help ensure that your pull request is accepted:
 
-Requirements for a GitHub issue to get consensus:
+  - Does the contribution alter the behavior of features or components in a way that it may break previous users’ programs and setups? If yes, there needs to be a discussion and agreement that this change is desirable.
+  - Does the contribution conceptually fit well into TiDB? Is it too much of a special case such that it makes things more complicated for the common case, or bloats the abstractions / APIs?
+  - Does the contribution make a big impact on TiDB's build time?
+  - Does your contribution affect any documentation? If yes, you should add/change proper documentation. 
+  - If there are any new dependencies, are they under active maintainences? What are their licenses?
```
```suggestion
* Does the contribution alter the behavior of features or components in a way that it may break previous users' programs and setups? If yes, there needs to be a discussion and agreement that this change is desirable.
* Does the contribution conceptually fit well into TiDB? Is it too much of a special case such that it makes things more complicated for the common case, or bloats the abstractions/APIs?
* Does the contribution make a big impact on TiDB's build time?
* Does your contribution affect any documentation? If yes, you should add/change proper documentation. 
* If there are any new dependencies, are they under active maintenances? What are their licenses?
```

tisonkun commented on src/contribute-to-tidb/contribute-code.md:
```diff
@@ -1,88 +1,53 @@
+Each feature or bugfix should be addressed by a single pull request, and for each pull request there may be several commits. In particular:
 
-Contributors can expect to get a first reaction from a committer within a few days after opening the issue. If an issue doesn’t get any attention, we recommend reaching out to the [TiDB Internals forum](https://internals.tidb.io/). Note that the community sometimes does not have the capacity to accept all incoming contributions.
+  - Do *not* fix more than one issue in the same commit (except, of course, if one code change fixes all of them).
+  - Do *not* do cosmetic changes to unrelated code in the same commit as some feature/bugfix.
```
```suggestion
* Do *not* fix more than one issues in the same commit (except, of course, if one code change fixes all of them).
* Do *not* do cosmetic changes to unrelated code in the same commit as some feature/bugfix.
```

tisonkun commented on src/contribute-to-tidb/contribute-code.md:
```diff
@@ -1,88 +1,53 @@
+## Waiting for review
 
-### 2. Implement your change
+To begin with, please be patient! There are many more people submitting pull requests than there are people capable of reviewing your pull request. Getting your pull request reviewed requires a reviewer to have the spare time and motivation to look at your pull request. If your pull request has not received any notice from reviewers (i.e., no comment made) for some time, you can ping the reviewer ...
```
```suggestion
To begin with, please be patient! There are many more people submitting pull requests than there are people capable of reviewing your pull request. Getting your pull request reviewed requires a reviewer to have the spare time and motivation to look at your pull request. If your pull request has not received any notice from reviewers (i.e., no comment made) for some time, you can ping the reviewers and assignees, or take it to [internal.tidb.io](https://internals.tidb.io) for more attention.
```

tisonkun commented on src/contribute-to-tidb/contribute-code.md:
```diff
@@ -1,88 +1,53 @@
-### 4. Merge change
-
-The code will be merged by a committer of TiDB once the review is finished. The related GitHub issue will be closed afterwards.
+When someone does manage to find the time to look at your pull request they will most likely make comments about how it can be improved (don’t worry, even committers/maintainers have their pull requests sent back to them for changes). It is then expected that you update your pull request to address these comments, and the review process will thus iterate until a satisfactory solution has emerged.
```
```suggestion
When someone does manage to find the time to look at your pull request, they will most likely make comments about how it can be improved (don't worry, even committers/maintainers have their pull requests sent back to them for changes). It is then expected that you update your pull request to address these comments, and the review process will thus iterate until a satisfactory solution has emerged.
```

tisonkun reviewed (commented):
Thanks for refactoring contribute code section. Comments inline.

tisonkun commented:
> * still no clear TiDB branch strategy.
> * still no commit message rule consensus
> * still no TiDB roadmap or feature map
> * still no experts list
> * still no clear label strategy

I think you can create issues for them so that we don't lost trace.

tisonkun reviewed (commented):
Also I think you can also take a look at https://github.com/pingcap/tidb-dev-guide/blob/master/src/get-started/commit-code-and-submit-a-pull-request.md to keep we express consistently.

zhangyangyu commented on src/contribute-to-tidb/contribute-code.md:
```diff
@@ -1,88 +1,53 @@
+Contributing to TiDB does *not* start with opening a pull request. We expect contributors to reach out to us first to discuss the overall approach together. Without consensus with the TiDB committers, contributions might require substantial rework or will not be reviewed. So please [create a GitHub issue](open-an-issue.md), discuss under an existing issue or create a topic on the [internal.tidb.i ...
 
-## Looking for what to contribute
+For newcomers, you can check the [starter issues](https://github.com/pingcap/tidb/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22), which are annotated with a "good first issue" label. These are issues suitable for new contributors to work with and won't take long to fix. But because the label is typically added at triage time it can turn out to be inaccurate, so do feel free to leav ...
```
Nice. Better.

zhangyangyu commented on src/contribute-to-tidb/contribute-code.md:
```diff
@@ -1,88 +1,53 @@
+For newcomers, you can check the [starter issues](https://github.com/pingcap/tidb/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22), which are annotated with a "good first issue" label. These are issues suitable for new contributors to work with and won't take long to fix. But because the label is typically added at triage time it can turn out to be inaccurate, so do feel free to leav ...
 
-If you have a good idea for the contribution, you can proceed to the code contribution process. If you are looking for what you could contribute, you can browse open [GitHub issues](https://github.com/pingcap/tidb/issues) of TiDB, which are not assigned, and then follow the code contribution process. If you are very new to the TiDB project and want to learn about it and its contribution process,  ...
+To get your change merged you need to sign the [CLA](https://cla-assistant.io/pingcap/tidb?pullRequest=5567) to grant PingCAP ownership of your code.
```
Actually I prefer to tell the user the CLA is a must. We just tell them it's really needed. If they really want to know detail, they should read the full CLA document.

zhangyangyu commented:
Addressed the comments. @tisonkun Please review again. 

I've read the get-started part. I think these two chapter don't have much in common. get-started is a detailed, step-by-step git guide, and this chapter focuses on more.

tisonkun reviewed (approved):
LGTM. Thanks for your work! It's very instructive.
